### ISSUE TITLE
AKAZE: 16-bit image support?

### ISSUE THREAD

**Issue by opencv-pushbot** [opencv/opencv#5061]
Transferred from http://code.opencv.org/issues/4467
```
|| Francesco Pierfederici on 2015-07-07 11:17
|| Priority: Normal
|| Affected: branch 'master' (3.0-dev)
|| Category: None
|| Tracker: Bug
|| Difficulty: 
|| PR: 
|| Platform: x86 / Mac OSX
```

AKAZE: 16-bit image support?
-----------
```
Hi there, I am a totally newbie here and so I might be asking for something that does not make sense.

I have noticed (in OpenCV 3.0.0 on Mac OS X 10.10.4) that due to this piece of code in akaze.cpp:171

if (img.type() != CV_8UC1)
    cvtColor(image, img, COLOR_BGR2GRAY);

running the AKAZE detectAndCompute() on 16-bit greyscale images will cause an exception:

    color.cpp:7564: error: (-215) scn == 3 || scn == 4 in function cvtColor

since the input image is already in greyscale, just not 8-bit/pixel. Would it make sense to change the above code to 

if (img.type() != CV_8UC1 && img.type() != CV_16UC1)
    cvtColor(image, img, COLOR_BGR2GRAY);

to allow processing of 16-bit greyscale images as well?
```

History
-------

**Comment by StevenPuttemans** [comment]
You should be very careful with adding that exception. There is a fair chance that the 16-bit image will meet a function further down the line that is not able to handle this format.

**Comment by nitsch** [comment]
As of version 3.1.0, multi-channel images are handled properly by AKAZE (internally converting them to grayscale). Accepted depths are 32F, 8U, 16U.

Looks like this issue can be closed.

**Comment by mshabunin** [comment]
Fixed here: #5658
